### PR TITLE
migrate effects-examples to use dune

### PR DESCRIPTION
**Pull request by shubhamkumar13**
Migrate aio from ocamlbuild to dune #19

### CODE REVIEW

**Comment by shubhamkumar13** [comment]
@kayceesrk I only found `aio/..` using `ocamlbuild` so I ported it to dune also I wasn't sure if all the commands Makefiles using `ocamlc/ocamlopt` also need to be ported to dune. If so do let me know.

**Comment by kayceesrk** [comment]
Let's port all the builds to dune.

File: aio/Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 all:
-	ocamlbuild -package unix echo.native
+	@dune exec ./echo.bc
```
Comment (kayceesrk) [inline aio/Makefile:2]: Is this executing the command? I don't think you should execute the command. `echo` server blocks and will not return.

File: aio/dune [hunk 2]
```diff
@@ -0,0 +1,5 @@
+(executables
+ (names echo)
+ (libraries unix)
+ (modules aio echo)
+ (modes byte exe))
```
Comment (kayceesrk) [inline aio/dune:5]: We don't need the bytecode mode. Observe that `ocamlbuild` command only builds the native version.

File: aio/echo.ml [hunk 3]
```diff
@@ -22,7 +22,7 @@
  * [2] https://github.com/ocamllabs/opam-repo-dev
  *)
 
-open Printexc
+(*open Printexc*)
```
Comment (kayceesrk) [inline aio/echo.ml:25]: Deleting unused code is always better than commenting them.

**Comment by kayceesrk** [comment]
The CI is failing for both the pipelines. Did you build it locally? and did it succeed locally?

File: aio/dune (hunk 2 above)
Comment (shubhamkumar13) [reply aio/dune]: Saw that later in the makefile.

**Comment by shubhamkumar13** [comment]
> The CI is failing for both the pipelines. Did you build it locally? and did it succeed locally?

Yes I ran it locally and it did succeed to build it locally

**Comment by kayceesrk** [comment]
> Yes I ran it locally and it did succeed to build it locally

You should investigate CI failure. The MacOS build is known to succeed. See #21.

File: Makefile [hunk 4]
```diff
@@ -4,29 +4,46 @@ EXE := concurrent.exe state.exe ref.exe transaction.exe echo.exe delimcc.exe \
...
 
 concurrent.exe: sched.mli sched.ml concurrent.ml
-	ocamlopt -o concurrent.exe sched.mli sched.ml concurrent.ml
+	@dune build concurrent.exe
 
 echo.exe: aio/aio.mli aio/aio.ml aio/echo.ml
-	$(MAKE) -C aio
-	cp aio/echo.native echo.exe
+	dune build aio/echo.exe
+	cp _build/default/aio/echo.exe .
```
Comment (kayceesrk) [inline Makefile:17]: None of this `cp` commands is necessary as the executables can be run as `dune run aio/echo.exe` for example.

File: eratosthenes.ml [hunk 5]
```diff
@@ -17,6 +17,7 @@ effect Yield : unit
 let yield () = perform Yield
 
 (** Communication primitives **)
+
```
Comment (kayceesrk) [inline eratosthenes.ml:20]: Is this change necessary?

File: mvar/MVar.ml [hunk 6]
```diff
@@ -32,7 +32,7 @@ module Make (S : SCHED) : S = struct
 
   let put v mv =
     match !mv with
-    | Full (v', q) -> suspend (fun k -> Queue.push (v,k) q)
+    | Full (v, q) -> suspend (fun k -> Queue.push (v,k) q)
```
Comment (kayceesrk) [inline mvar/MVar.ml:35]: Are you sure that this change is correct? This changes the semantics.

Comment (kayceesrk) [reply mvar/MVar.ml:35]: On further thoughts, this change is clearly incorrect. Please revert this change.

File: mvar/MVar.mli [hunk 7]
```diff
@@ -20,14 +20,17 @@ end
 
 module type SCHED = sig
   type 'a cont
+
```
Comment (kayceesrk) [inline mvar/MVar.mli:23]: Is this and further spacing edits necessary? The additional space messes up documentation generation. The documentation comments `(** ... *)` must be attached to specific items in the `mli` file for it to work correctly.

File: mvar/sched.mli [hunk 8]
```diff
@@ -1,21 +1,27 @@
 type 'a cont
+
```
Comment (kayceesrk) [inline mvar/sched.mli:2]: Do not add extra space.

File: reify_reflect.ml [hunk 9]
```diff
@@ -77,21 +77,21 @@ let state_example () =
 (* Using the exception monad *)
 let exception_example () =
   Printf.printf "Raising an exception\n";
-  raise (Failure "An error!");
+  (*raise (Failure "An error!");*)
```
Comment (kayceesrk) [inline reify_reflect.ml:80]: Do not change semantics.

File: transaction.ml [hunk 10]
```diff
@@ -42,7 +42,7 @@ let () = atomically (fun () ->
     r := 20;
     r := 21;
     printf "T1: Before abort %d\n" (!r);
-    raise (Res !r);
+    (* raise (Res !r); *)
```
Comment (kayceesrk) [inline transaction.ml:45]: Do not change semantics.
